### PR TITLE
hash cost for users decreased

### PR DESCRIPTION
**Pull request by mvr320**
Symfony does advice 'Sodium' which could be used as alternative, or we
could set all values lower and advice a reasonable amount of hashes per
time.

Also added a config check to verify if hashing time is reasonable.

Locally 7 is to simple, but I think thats something someone should propose ranges for..

This is the start for https://github.com/DOMjudge/domjudge/issues/936

### CODE REVIEW

**Comment by meisterT** [comment]
If you generate passwords for 200 users, how long does it take before/after your change?

File: webapp/src/Controller/Jury/ConfigController.php [hunk 1]
```diff
@@ -16,6 +16,7 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
```
Comment (nickygerritsen) [inline webapp/src/Controller/Jury/ConfigController.php:19]: Seems like this is not needed (anymore)?

File: webapp/src/Service/CheckConfigService.php [hunk 2]
```diff
@@ -337,6 +346,36 @@ public function checkTmpdirWritable()
...
 
+    public function checkHashTime()
+    {
+        $tmpuser = new User();
+        $counter = 0;
+        $time_start = microtime(true);
+        do {
+            $plainPassword = 'ryanpass';
+            $encoded = $this->passwordEncoder->encodePassword($tmpuser, $plainPassword);
+            // $user->setPassword($encoded);
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:357]: This comment can be removed

File: webapp/src/Service/CheckConfigService.php [hunk 3]
```diff
@@ -337,6 +346,36 @@ public function checkTmpdirWritable()
...
+            $time_end = microtime(true);
+            $counter++;
+        } while ( ($time_end - $time_start) < 4);
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:360]: This 4 is the number of seconds right? Maybe we should lower it a bit if that's the case? I guess that we can also determine a sane value in 2 seconds.

**Comment by mvr320** [comment]
> 
> 
> If you generate passwords for 200 users, how long does it take before/after your change?

I can do 429 user hashes in 4s if the logic of the config page is indeed correct.

**Comment by meisterT** [comment]
And with the old settings?

File: webapp/src/Service/CheckConfigService.php [hunk 4]
```diff
@@ -337,6 +346,36 @@ public function checkTmpdirWritable()
...
+
+        $length_suggestion = 'When you have a contest duration more than a week increase the password length.';
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:362]: This is not related to the hashing algo right? But more a general thing.

File: webapp/src/Service/CheckConfigService.php [hunk 5]
```diff
@@ -337,6 +346,36 @@ public function checkTmpdirWritable()
...
+        if ($counter>300) {
+            return ['caption' => 'User password hashing',
+                'result' => 'W',
+                'desc' => sprintf('Hashing is to simple for small sized contests (Did %d hashes). %s',
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:366]: `s/to/too`
Also maybe we need to add somewhere in the docs how to increase / lower the hashing cost?

File: webapp/src/Service/CheckConfigService.php [hunk 6]
```diff
@@ -337,6 +346,36 @@ public function checkTmpdirWritable()
...
+                                  $counter, $length_suggestion)];
+        } elseif ($counter<100) {
+            return ['caption' => 'User password hashing',
+                'result' => 'W',
+                'desc' => sprintf('Hashing is to expensive for medium sized contests  (Did %d hashes). %s',
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:371]: s/to/too`

**Review by nickygerritsen — CHANGES_REQUESTED** [review]

File: webapp/src/Service/CheckConfigService.php (hunk 4 above)
Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:362]: Yes and no, if you want more secure passwords the first thing which should be increased is the length (citing thijs) and not the hashing cost. I think i'll remove it and make something for that specific usecase?

File: webapp/src/Service/CheckConfigService.php (hunk 3 above)
Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:360]: I now delay the page for 4s, will change it to 2s

File: webapp/src/Service/CheckConfigService.php (hunk 2 above)
Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:357]: Thats something I wanted to ask you, I assume the encodePassword will already do the cost (so its not a promise to the DB?)

File: webapp/src/Service/CheckConfigService.php (hunk 4 above)
Comment (nickygerritsen) [reply webapp/src/Service/CheckConfigService.php:362]: True. We can't really check the length of passwords, so maybe just something we should put somewhere indeed as a note.

File: webapp/src/Service/CheckConfigService.php (hunk 3 above)
Comment (nickygerritsen) [reply webapp/src/Service/CheckConfigService.php:360]: Sounds good.

File: webapp/src/Service/CheckConfigService.php (hunk 2 above)
Comment (nickygerritsen) [reply webapp/src/Service/CheckConfigService.php:357]: Correct.

File: webapp/src/Controller/Jury/ConfigController.php (hunk 1 above)
Comment (mvr320) [reply webapp/src/Controller/Jury/ConfigController.php:19]: When I remove it things break, its used at L77

Comment (nickygerritsen) [reply webapp/src/Controller/Jury/ConfigController.php:19]: Line 77 is a comment. You are looking at the wrong file I think

Comment (mvr320) [reply webapp/src/Controller/Jury/ConfigController.php:19]: You think correctly.

File: webapp/src/Service/CheckConfigService.php (hunk 5 above)
Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:366]: Done

File: webapp/src/Service/CheckConfigService.php [hunk 7]
```diff
@@ -60,14 +61,20 @@ class CheckConfigService
      */
     protected $debug;
 
+    /**
+     * @var $passwordEncoder
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:65]: this @var is quite useless.

File: webapp/src/Service/CheckConfigService.php [hunk 8]
```diff
@@ -60,14 +61,20 @@ class CheckConfigService
...
+     */
+    protected $passwordEncoder;
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:67]:
```suggestion
    protected UserPasswordEncoderInterface $passwordEncoder;
```

File: webapp/src/Service/CheckConfigService.php (hunk 7 above)
Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:65]: All of these "@Var" should be removed right? (in that section)

**Review by nickygerritsen — APPROVED** [review]

**Pull request merged**